model/adjuster/span_hash_deduper.go

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2019 The Jaeger Authors.
+// Copyright (c) 2017 Uber Technologies, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"github.com/jaegertracing/jaeger/model"
+)
+
+// DedupeBySpanHash returns an adjuster that removes all but one span with the same hashcode.
+// This is useful for when spans are duplicated in archival storage, as happens with
+// ElasticSearch archival.
+func DedupeBySpanHash() Adjuster {
+	return Func(func(trace *model.Trace) (*model.Trace, error) {
+		deduper := &spanHashDeduper{trace: trace}
+		deduper.groupSpansByHash()
+		deduper.dedupeSpansByHash()
+		return deduper.trace, nil
+	})
+}
+
+type spanHashDeduper struct {
+	trace       *model.Trace
+	spansByHash map[uint64][]*model.Span
+}
+
+func (d *spanHashDeduper) groupSpansByHash() {
+	spansByHash := make(map[uint64][]*model.Span)
+	for _, span := range d.trace.Spans {
+		hash, _ := model.HashCode(span)
```

Comment:
I recall there was some issue in the past with hash function requiring proper ordering of tags or something.

Reply:
Yes, this is still the case:
```
func (s *Span) Hash(w io.Writer) (err error) {
	// gob is not the most efficient way, but it ensures we don't miss any fields.
	// See BenchmarkSpanHash in span_test.go
	enc := gob.NewEncoder(w)
	return enc.Encode(s)
}
```
You can only compare hashes if the tags are properly sorted. Surprisingly, we don't have an explicit adjuster for that, but some of that sorting happens in ip_tag.go and sort_log_fields.go adjusters. Perhaps it would make sense to pull the sorting explicitly into SortTagsAndLogs adjuster and makes sure it is called before `DedupeBySpanHash` adjuster.

Reply:
OK, now we have SortTagsAndLogFields adjuster